src/lib.rs

```diff
@@ -251,42 +255,75 @@ impl EndpointHeader {
             let prefetchable = bar.get_bit(3);
             let address = bar.get_bits(4..32) << 4;
 
-            // TODO: if the bar is 64-bits, do we need to do this on both BARs?
-            let size = unsafe {
-                access.write(self.0, offset, 0xffffffff);
-                let mut readback = access.read(self.0, offset);
-                access.write(self.0, offset, address);
-
-                /*
-                 * If the entire readback value is zero, the BAR is not implemented, so we return `None`.
-                 */
-                if readback == 0x0 {
-                    return None;
-                }
-
-                readback.set_bits(0..4, 0);
-                1 << readback.trailing_zeros()
-            };
-
-            match bar.get_bits(1..3) {
-                0b00 => Some(Bar::Memory32 {
-                    address,
-                    size,
-                    prefetchable,
-                }),
+            match bar.get_bits(1..3)
+            {
+                0b00 => {
+                    let size = unsafe {
+                        access.write(self.0, offset, 0xfffffff0);
+                        let mut readback = access.read(self.0, offset).get_bits(4..32);
+                        access.write(self.0, offset, address);
+
+                        if readback == 0x0 {
```

Comment:
Might be nice to retain the comment about zero length being speced to mean a BAR is not implemented.